File: mobile/src/bindings/plan.ts
import { invoke } from "@tauri-apps/api/core";
import type { PlanView } from "./types";

export const plan = {
  async add(name: string): Promise<void> {
    await invoke("plan", { command: { action: "Add", name } });
  },

  async remove(name: string): Promise<void> {
    await invoke("plan", { command: { action: "Remove", name } });
  },

  async view(name: string): Promise<PlanView> {
    return await invoke("plan", { command: { action: "View", name } });
  },

  async list(): Promise<string[]> {
    return await invoke("plan", { command: { action: "List" } });
  },

  async rename(name: string, newName: string): Promise<void> {
    await invoke("plan", { command: { action: "Rename", name, new_name: newName } });
  },

  async assign(
    planName: string,
    indexName: string,
    daypartName: string,
    mealName: string,
  ): Promise<void> {
    await invoke("plan", {
      command: {
        action: "Assign",
        planname: planName,
        indexname: indexName,
        daypartname: daypartName,
        mealname: mealName,
      },
    });
  },

  async unassign(
    planName: string,
    indexName: string,
    daypartName: string,
  ): Promise<void> {
    await invoke("plan", {
      command: {
        action: "Unassign",
        planname: planName,
        indexname: indexName,
        daypartname: daypartName,
      },
    });
  },

  async fill(planName: string, days: number = 7): Promise<void> {
    await invoke("plan", { command: { action: "Fill", planname: planName, days } });
  },
};

File: mobile/src-tauri/src/commands/plan.rs
use foody_core::app::App;
use serde::Deserialize;
use tauri::State;

use crate::commands::serialize_json;

#[derive(Deserialize)]
#[serde(tag = "action")]
pub enum PlanCommand {
    Add { name: String },
    Remove { name: String },
    View { name: String },
    List,
    Rename { name: String, new_name: String },
    Assign {
        planname: String,
        indexname: String,
        daypartname: String,
        mealname: String,
    },
    Unassign {
        planname: String,
        indexname: String,
        daypartname: String,
    },
    Fill { planname: String, days: Option<i64> },
}

#[tauri::command]
pub async fn plan(
    state: State<'_, App>,
    command: PlanCommand,
) -> Result<serde_json::Value, String> {
    match command {
        PlanCommand::Add { name } => serialize_json(state.add_plan(&name)).await,
        PlanCommand::Remove { name } => serialize_json(state.remove_plan(&name)).await,
        PlanCommand::View { name } => serialize_json(state.view_plan(&name)).await,
        PlanCommand::List => serialize_json(state.list_plans()).await,
        PlanCommand::Rename { name, new_name } => {
            serialize_json(state.rename_plan(&name, &new_name)).await
        }
        PlanCommand::Assign {
            planname,
            indexname,
            daypartname,
            mealname,
        } => {
            serialize_json(
                state.assign_plan(&planname, &indexname, &daypartname, &mealname),
            )
            .await
        }
        PlanCommand::Unassign {
            planname,
            indexname,
            daypartname,
        } => {
            serialize_json(state.unassign_plan(&planname, &indexname, &daypartname)).await
        }
        PlanCommand::Fill { planname, days } => {
            serialize_json(state.fill_plan(&planname, days.unwrap_or(7) as usize)).await
        }
    }
}
